# Table of contents

1. [Las Probabilidades de los equipos de la CONMEBOL para clasificar a La Copa Mundial de Fútbol Qatar 2022](#introduccion)

2. [Generando el Espacio Muestral](#espacio)
    
    2.1 [Resultados por jornada](#jornada)
    
    2.2 [Puntos por partido disputado](#puntos)
    
    2.3 [Procesamiento](#procesamiento)
    
3. [Análisis de los partidos](#analisis)
    
    3.1 [¿Si Ecuador pierde todos los partidos puede clasificarse para el mundial sin repechaje?](#p1)
    
    3.2 [¿Con un empate Ecuador podría quedarse fuera del mundial sin opción a repechaje?](#p2)
    
    3.3 [¿Si Ecuador gana un partido o empatara los tres partidos podría quedarse fuera del mundial?](#p3)


## Las Probabilidades de los equipos de la CONMEBOL para clasificar a La Copa Mundial de Fútbol Qatar 2022 <a name="introduccion"></a>

Los 4 primeros equipos que obtengan el puntaje más alto de los 10 equipo de la CONMEBOL clasificarían directamente a La Copa Mundial de Fútbol Qatar 2022. El quinto equipo con puntaje más alto competiría en un repechaje por una plaza para el mundial. Actualmente, faltan por disputar 3 jornadas de 5 cinco partidos cada uno, para que acabe la eliminatoria. A continuación, se detallan los partidos por jornada:

* Jornada 16: Uruguay vs Venezuela, Brasil vs Paraguay, Perú vs Ecuador, Bolivia vs Chile, Argentina vs Colombia.

* Jornada 17: Argentina vs Venezuela, Colombia vs Bolivia, Paraguay vs Ecuador, Brasil vs Chile, Uruguay vs Perú.

* Jornada 18: Perú vs Paraguay, Ecuador vs Argentina, Venezuela vs Colombia, Chile vs Uruguay, Bolivia vs Brasil.

En este trabajo se pretende obtener las probabilidades que tienen los equipos de la CONMEBOL para clasificar al mundial de fútbol. Para esto se debe contar con el conjunto de todos los resultados posibles de las jornadas 16, 17 y 18. 

El conjunto de todos los resultados posibles se denomina espacio muestral y se suele designar con la letra $\Omega$.

Para explicar cómo se generará el espacio muestral, se escogerán dos partidos de la jornada 16: Perú vs Ecuador y Bolivia vs Chile. A continuación, se detallan los escenarios posibles de estos dos partidos:



$$\Omega = \{\{Si~Perú~gana,~Ecuador~pierde;~Si~Bolivia~gana,~Chile~pierde\}, \{Si~Perú~gana,~Ecuador~pierde; ~Si~Bolivia~pierde,~Chile~gana\},\{Si~Perú~gana,~Ecuador~pierde;~Si~Bolivia~empata,~empata~Chile\},\{Si~Perú~pierde,~Ecuador~gana;~Bolivia~gana,~Chile~pierde\},\{Si~Perú~pierde,~Ecuador~gana;~Si~Bolivia~pierde,~Chile~gana\},\{Si~Perú~pierde,~Ecuador~gana;~Si~Bolivia~empata,~Chile~empata\},\{Si~Perú~empata,~Ecuador~empata;~Si~Bolivia~gana,~Chile~pierde\},\{Si~Perú~empata,~Ecuador~empata;~Si~Bolivia~pierde,~Chile~gana\},\{Si~Perú~empata,~Ecuador~empata;~Si~Bolivia~empata,~Chile~empata\}\}
$$


Como se puede observar, $\Omega$ en dos partidos tendría $3\cdot3 = 3^2 = 9$ escenarios posibles. En tres partidos, $\Omega$ tendría $3\cdot3\cdot3 = 3^3 = 27$ escenarios posibles. Por lo tanto, la fórmula para calcular $\Omega$ o el espacio muestral de una jornada de la sería:

$$
\Omega~de~una~jornada = resultados~posibles^{número~de~partidos}
$$


Las jornadas 16, 17 y 18 tienen 5 partidos con 3 posibles resultados por partido, por lo que el $\Omega$ de cada jornada tendrían los siguientes sucesos:

$$
\Omega~de~la~jornada~16 = 3^{5} = 243 ~ escenarios
$$

Por lo tanto, el espacio muestral de las 3 jornadas restantes sería:

$$
\Omega~de~la~jornada~16 \cdot \Omega~de~la~jornada~17 \cdot \Omega~de~la~jornada~16 = 243 \cdot 243 \cdot 243 = 14.348.907 sucesos
$$


## Generando el Espacio Muestral <a name="espacio"></a>

Para generar el espacio muestral de las jornadas 16, 17 y 18 se utilizarán las funciones que se detallan a continuación.

### Resultados por jornada <a name="jornada"></a>

La función `get_results` retorna los resultados de ganar, perder y empatar de 5 partidos de una jornada.

In [23]:
def get_results():
    # 0 gana el local, 1 gana el visitante, 2 empate
    score = [0,1,2]

    results = []
    for sc1 in score:
        for sc2 in score:
            for sc3 in score:
                for sc4 in score:
                    for sc5 in score:
                        results.append([sc1,sc2, sc3, sc4, sc5])
    return results

In [24]:
results = get_results()
len(results)

243

Se comprueba que son 243 sucesos por jornada.

### Puntos por partido disputado <a name="puntos"></a>

La funcion `get_standings` establece 3 puntos al equipo ganador, 0 puntos al equipo perdedor y 1 puntos a los equipos que empaten.

In [25]:
def get_standings(resultsc, matchday):
    standings = []
    temp_standing = {"BR": 0, "AR": 0, "EC": 0, "UR": 0, "CL": 0, "PE": 0, "CH": 0, "BL": 0, "PG": 0, "VZ": 0, "M1": "", "M2": "", "M3": "", "M4": ""}
    for result in resultsc:
        match0 = matchday[0]
        match0_result = result[0]

        match1 = matchday[1]
        match1_result = result[1]

        match2 = matchday[2]
        match2_result = result[2]

        match3 = matchday[3]
        match3_result = result[3]

        match4 = matchday[4]
        match4_result = result[4]

        temp_match0 = ""
        temp_match1 = ""
        temp_match2 = ""
        temp_match3 = ""
        temp_match4 = ""

        if match0_result == 0:
            temp_standing[match0[0]] = 3
            temp_standing[match0[1]] = 0
            temp_match0 += match0[0] + " (3) " + ' - '+ match0[1] + " (0)"
        elif match0_result == 1:
            temp_standing[match0[0]] = 0
            temp_standing[match0[1]] = 3
            temp_match0 += match0[0] + " (0) " + ' - '+ match0[1] + " (3)"
        else:
            temp_standing[match0[0]] = 1
            temp_standing[match0[1]] = 1
            temp_match0 += match0[0] + " (1) " + ' - '+ match0[1] + " (1)"

        if match1_result == 0:
            temp_standing[match1[0]] = 3
            temp_standing[match1[1]] = 0
            temp_match1 += match1[0] + " (3) " + ' - '+ match1[1] + " (0)"
        elif match1_result == 1:
            temp_standing[match1[0]] = 0
            temp_standing[match1[1]] = 3
            temp_match1 += match1[0] + " (0) " + ' - '+ match1[1] + " (3)"
        else:
            temp_standing[match1[0]] = 1
            temp_standing[match1[1]] = 1
            temp_match1 += match1[0] + " (1) " + ' - '+ match1[1] + " (1)"

        if match2_result == 0:
            temp_standing[match2[0]] = 3
            temp_standing[match2[1]] = 0
            temp_match2 += match2[0] + " (3) " + ' - '+ match2[1] + " (0)"
        elif match2_result == 1:
            temp_standing[match2[0]] = 0
            temp_standing[match2[1]] = 3
            temp_match2 += match2[0] + " (0) " + ' - '+ match2[1] + " (3)"
        else:
            temp_standing[match2[0]] = 1
            temp_standing[match2[1]] = 1
            temp_match2 += match2[0] + " (1) " + ' - '+ match2[1] + " (1)"

        if match3_result == 0:
            temp_standing[match3[0]] = 3
            temp_standing[match3[1]] = 0
            temp_match3 += match3[0] + " (3) " + ' - '+ match3[1] + " (0)"
        elif match3_result == 1:
            temp_standing[match3[0]] = 0
            temp_standing[match3[1]] = 3
            temp_match3 += match3[0] + " (0) " + ' - '+ match3[1] + " (3)"
        else:
            temp_standing[match3[0]] = 1
            temp_standing[match3[1]] = 1
            temp_match3 += match3[0] + " (1) " + ' - '+ match3[1] + " (1)"


        if match4_result == 0:
            temp_standing[match4[0]] = 3
            temp_standing[match4[1]] = 0
            temp_match4 += match4[0] + " (3) " + ' - '+ match4[1] + " (0)"
        elif match4_result == 1:
            temp_standing[match4[0]] = 0
            temp_standing[match4[1]] = 3
            temp_match4 += match4[0] + " (0) " + ' - '+ match4[1] + " (3)"
        else:
            temp_standing[match4[0]] = 1
            temp_standing[match4[1]] = 1
            temp_match4 += match4[0] + " (1) " + ' - '+ match4[1] + " (1)"
        
        temp_standing["M1"] = temp_match0
        temp_standing["M2"] = temp_match1
        temp_standing["M3"] = temp_match2
        temp_standing["M4"] = temp_match3
        temp_standing["M5"] = temp_match4

        standings.append(temp_standing)
        temp_standing = {"BR": 0, "AR": 0, "EC": 0, "UR": 0, "CL": 0, "PE": 0, "CH": 0, "BL": 0, "PG": 0, "VZ": 0, "M1": "", "M2": "", "M3": "", "M4": ""}
    return standings

A continuación, se establecen los partidos de las jornadas 16, 17 y 18 en las variables `matchday16`, `matchday17` y `matchday18`

In [27]:
matchday16 = [["UR", "VZ"],["BR", "PG"],["PE", "EC"],["BL", "CH"],["AR", "CL"]]
matchday17 = [["AR", "VZ"],["CL", "BL"],["PG", "EC"],["BR", "CH"],["UR", "PE"]]
matchday18 = [["PE", "PG"],["EC", "AR"],["VZ", "CL"],["CH", "UR"],["BL", "BR"]]

Las variables `standings1`, `standings2` y `standings3` contienen los posibles puntos obtenidos por los equipos en las jornadas 16, 17 y 18.

In [28]:
standings1 = get_standings(results, matchday16)
standings2 = get_standings(results, matchday17)
standings3 = get_standings(results, matchday18)

len(standings1) * len(standings2) * len(standings3)

14348907

Se comprueba que el $\Omega$ de las tres jornadas es 14348907 sucesos.

### Procesamiento <a name="procesamiento"></a>

A continuación, se importa la librería `pandas`, la cual nos facilitará consultar los resultados de las jornadas.

In [29]:
import pandas as pd

En la variable `df` de tipo `DataFrame`, se almacenaran los puntos obtenidos en cada escenario.

In [30]:
df = pd.DataFrame(
    columns=['J16', 'J17','J18',
    'BR', 'AR', 'EC', 'CL', 'PE', 'CH', 'UR', 'BL', 'PG', 'VZ',
    'BR_POS', 'AR_POS', 'EC_POS', 'CL_POS', 'PE_POS', 'CH_POS', 'UR_POS', 'BL_POS', 'PG_POS', 'VZ_POS'])
# df = df.astype({'J16': int, 'J17': int,'J18': int, 'BR': int, 'AR': int, 'EC': int, 'CL': int, 'PE': int, 'CH':int, 'UR': int, 'BL': int, 'PG': int, 'VZ': int})

A continuación, se detallan los campos del dataset:

* J16: contendrá el índice que identifica la combinación de partidos de la jornada 16.
* J17: contendrá el índice que identifica la combinación de partidos de la jornada 17.
* J18: contendrá el índice que identifica la combinación de partidos de la jornada 18.
* BR, AR, EC, CL, PE, CH, UR, BL, PG, VZ: los posibles puntos de los países.

A continuacion, se establece el puntaje actual que tienen las selecciones

In [183]:
current_standing = {"BR": 36, "AR": 32, "EC": 24, "UR": 19, "CL": 17, "PE": 20, "CH": 16, "BL": 15, "PG": 13, "VZ": 10}

En las siguientes sentencias, se crea el espacio muestral de las 3 jornadas:

In [54]:
dictinary_list = []

for j1, st1 in enumerate(standings1):
    for j2, st2 in enumerate(standings2):
        for j3, st3 in enumerate(standings3):
            sc_br = st1['BR']+st2['BR']+st3['BR'] + current_standing['BR']
            sc_ar = st1['AR']+st2['AR']+st3['AR'] + current_standing['AR']
            sc_ec = st1['EC']+st2['EC']+st3['EC'] + current_standing['EC']
            sc_cl = st1['CL']+st2['CL']+st3['CL'] + current_standing['CL']
            sc_pe = st1['PE']+st2['PE']+st3['PE'] + current_standing['PE']
            sc_ch = st1['CH']+st2['CH']+st3['CH'] + current_standing['CH']
            sc_ur = st1['UR']+st2['UR']+st3['UR'] + current_standing['UR']
            sc_bl = st1['BL']+st2['BL']+st3['BL'] + current_standing['BL']
            sc_pg = st1['PG']+st2['PG']+st3['PG'] + current_standing['PG']
            sc_vz = st1['VZ']+st2['VZ']+st3['VZ'] + current_standing['VZ']
            
            x = {"BR": sc_br, "AR": sc_ar, "EC": sc_ec, "CL": sc_cl, "PE": sc_pe, 
                 "CH": sc_ch, "UR": sc_ur, "BL": sc_bl, "PG": sc_pg, "VZ":sc_vz}
            
            r = {k: v for k, v in sorted(x.items(), key=lambda item: item[1], reverse=True)}
    
            pos_br = 0; pos_ar = 0; pos_ec = 0; pos_cl = 0; pos_pe = 0; 
            pos_ch = 0; pos_ur = 0; pos_bl = 0; pos_pg = 0; pos_vz = 0

            for pos, i in enumerate(r):
                if i == "BR":
                    pos_br = pos+1
                elif i == "AR":
                    pos_ar = pos+1
                elif i == "EC":
                    pos_ec = pos+1
                elif i == "CL":
                    pos_cl = pos+1
                elif i == "PE":
                    pos_pe = pos+1
                elif i == "CH":
                    pos_ch = pos+1
                elif i == "UR":
                    pos_ur = pos+1
                elif i == "BL":
                    pos_bl = pos+1
                elif i == "PG":
                    pos_pg = pos+1
                elif i == "VZ":
                    pos_vz = pos+1
            
            
            dictionary_data = {
                'J16': j1,
                'J17': j2,
                'J18': j3,
                'BR': sc_br,
                'AR': sc_ar,
                'EC': sc_ec,
                'CL': sc_cl,
                'PE': sc_pe,
                'CH': sc_ch,
                'UR': sc_ur,
                'BL': sc_bl,
                'PG': sc_pg,
                'VZ': sc_vz,
                'BR_POS': pos_br,
                'AR_POS': pos_ar,
                'EC_POS': pos_ec,
                'CL_POS': pos_cl,
                'PE_POS': pos_pe,
                'CH_POS': pos_ch,
                'UR_POS': pos_ur,
                'BL_POS': pos_bl,
                'PG_POS': pos_pg,
                'VZ_POS': pos_vz
                }

            dictinary_list.append(dictionary_data)

Luego, se crea un archivo `.csv` que contendra el espacio muestral con los puntos obtenidos.

In [55]:
df_final = pd.DataFrame.from_dict(dictinary_list)
df_final.to_csv('omega_score_position_v1.csv', index=False)

In [139]:
df_st1 = pd.DataFrame(standings1, columns=['M1', 'M2', 'M3', 'M4', 'M5'])
df_st2 = pd.DataFrame(standings2, columns=['M1', 'M2', 'M3', 'M4', 'M5'])
df_st3 = pd.DataFrame(standings3, columns=['M1', 'M2', 'M3', 'M4', 'M5'])

df_st1.to_csv('matchday16.csv', index=False)
df_st2.to_csv('matchday17.csv', index=False)
df_st3.to_csv('matchday18.csv', index=False)

In [56]:
df = pd.read_csv('omega_score_position_v1.csv')
df.head()

J16  J17  J18  BR  AR  EC  CL  PE  CH  UR  ...  BR_POS  AR_POS  EC_POS  \
0    0    0    0  42  38  27  20  26  19  25  ...       1       2       3   
1    0    0    1  45  38  27  20  26  19  25  ...       1       2       3   
2    0    0    2  43  38  27  20  26  19  25  ...       1       2       3   
3    0    0    3  42  38  27  20  26  16  28  ...       1       2       4   
4    0    0    4  45  38  27  20  26  16  28  ...       1       2       4   

   CL_POS  PE_POS  CH_POS  UR_POS  BL_POS  PG_POS  VZ_POS  
0       7       4       8       5       6       9      10  
1       6       4       7       5       8       9      10  
2       6       4       7       5       8       9      10  
3       7       5       8       3       6       9      10  
4       6       5       8       3       7       9      10  

[5 rows x 23 columns]

In [107]:
df_st1 = pd.read_csv('matchday16.csv')
df_st2 = pd.read_csv('matchday17.csv')
df_st3 = pd.read_csv('matchday18.csv')

## Análisis de los partidos <a name="analisis">

## ¿Si Ecuador pierde todos los partidos puede clasificarse para el mundial sin repechaje?<a name="p1">

In [156]:
result = df[(df_pts['EC'] == 24) & (df['EC_POS'] < 5)]
print("Probabilidad: ", len(result)/len(df))
print("Número de escenarios: ", len(result))

Probabilidad:  0.030590413611294574
Número de escenarios:  438939


Si, auquen la probabilidad es muy baja. Para que Ecuador clasifique para el mundial perdiendo todos los partidos deberían ocurrir los siguientes resultados:

In [169]:
df_st1.loc[result['J16']].head(1)

M1                M2                M3                M4  \
0  UR (3)  - VZ (0)  BR (3)  - PG (0)  PE (3)  - EC (0)  BL (3)  - CH (0)   

                 M5  
0  AR (3)  - CL (0)

In [154]:
df_st2.loc[result['J17']].head(1)

M1                M2                M3                M4  \
0  AR (3)  - VZ (0)  CL (3)  - BL (0)  PG (3)  - EC (0)  BR (3)  - CH (0)   

                 M5  
0  UR (3)  - PE (0)

In [170]:
df_st3.loc[result['J18']].head(1)

M1                M2                M3                M4  \
108  PE (0)  - PG (3)  EC (0)  - AR (3)  VZ (3)  - CL (0)  CH (3)  - UR (0)   

                   M5  
108  BL (3)  - BR (0)

*  Jornada 16: Uruguay gana a Venezuela, Brasil gana a Paraguay, Perú gana a Ecuador, Bolivia gana a Chile, Argentina gana a Colombia
*  Jornada 17: Argentina gana a Venezuela, Colombia gana a Bolivia, Paraguay gana a Ecuador, Brasil gana a Chile, Uruguay gana a Perú
*  Jornada 18: Paraguay gana a Perú, Argentina gana a Ecuador, Venezuela gana a Colombia, Chile gana a Uruguay, Bolivia gana a Brasil. 


## ¿Con un empate Ecuador podría quedarse fuera del mundial sin opción a repechaje?<a name="p2">

In [149]:
result = df[(df_pts['EC'] == 25) & (df['EC_POS'] > 5)]
print("Probabilidad: ", len(result)/len(df))
print("Número de escenarios: ", len(result))

Probabilidad:  0.00013548070246744229
Número de escenarios:  1944


Si, auquen la probabilidad es muy baja. Para que Ecuador empate y no se clasifique para el mundial deberían ocurrir los siguientes resultados:

In [145]:
df_st1.loc[result['J16']].head(1)

M1                M2                M3                M4  \
1  UR (3)  - VZ (0)  BR (3)  - PG (0)  PE (3)  - EC (0)  BL (3)  - CH (0)   

                 M5  
1  AR (0)  - CL (3)

In [142]:
df_st2.loc[result['J17']].head(1)

M1                M2                M3                M4  \
0  AR (3)  - VZ (0)  CL (3)  - BL (0)  PG (3)  - EC (0)  BR (3)  - CH (0)   

                 M5  
0  UR (3)  - PE (0)

In [146]:
df_st3.loc[result['J18']].head(1)

M1                M2                M3                M4  \
66  PE (3)  - PG (0)  EC (1)  - AR (1)  VZ (0)  - CL (3)  CH (0)  - UR (3)   

                  M5  
66  BL (3)  - BR (0)

* En la jornada 16: Uruguay gana a Venezuela, Brasil gana a Paraguay, Perú gana a Ecuador, Bolivia gana a Chile, Colombia gana a Argentina
* En la jornada 17: Argentina gana a Venezuela, Colombia gana a Bolivia, Paraguay gana a Ecuador, Brasil gana a Chile, Uruguay gana a Perú.
* En la jornada 18: Perú gana a Paraguay, Ecuador empate con Argentina, Colombia gana a Venezuela, Uruguay gana a Chile, Bolivia gana a Ecuador.

## ¿Si Ecuador gana un partido o empatara los tres partidos podría quedarse fuera del mundial?<a name="p3">

In [180]:
result = df[(df_pts['EC'] == 27) & (df['EC_POS'] > 4)]
print("Probabilidad: ", len(result)/len(df))
print("Número de escenarios: ", len(result))

Probabilidad:  0.0
Número de escenarios:  0


No, clasificaría directo al mundial, sin ir al repechaje.